### PR TITLE
Demonstrate && evaluation inconsistencies

### PR DESCRIPTION
**Pull request by ana-pantilie**
Pre-submit checklist:
- Branch
    - [ ] Tests are provided (if possible)
    - [ ] Commit sequence broadly makes sense
    - [ ] Key commits have useful messages
    - [ ] Changelog fragments have been written (if appropriate)
    - [ ] Relevant tickets are mentioned in commit messages
    - [ ] Formatting, PNG optimization, etc. are updated
- PR
    - [ ] (For external contributions) Corresponding issue exists and is linked in the description
    - [ ] Targeting master unless this is a cherry-pick backport
    - [ ] Self-reviewed the diff
    - [ ] Useful pull request description
    - [ ] Reviewer requested

### CODE REVIEW

File: plutus-tx-plugin/test/Budget/9.6/andWithoutGHCOpts.pir.golden [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+                (v : a -> a -> Bool)
+                (v : a -> a -> Bool)
+                (v : a -> a -> a)
+                (v : a -> a -> a) ->
+                 v)
+  in
+  \(x : integer) (y : integer) ->
+    let
+      !l : Bool = `<` {integer} `$fOrdInteger` x 3
+      !r : Bool = `<` {integer} `$fOrdInteger` y 3
```
Comment (ana-pantilie) [inline plutus-tx-plugin/test/Budget/9.6/andWithoutGHCOpts.pir.golden:80]: Same as in the "local" case above.

File: plutus-tx-plugin/test/Budget/9.6/andWithLocal.uplc.golden [hunk 2]
```diff
@@ -0,0 +1,12 @@
+program
+  1.1.0
+  ((\x y ->
+      (\lessThanInteger ->
+         (\l ->
+            (\r -> force (case l [(delay r), (delay (constr 1 []))]))
+              (lessThanInteger y 3))
+           (lessThanInteger x 3))
+        (\x y ->
+           force ifThenElse (lessThanInteger x y) (constr 0 []) (constr 1 [])))
```
Comment (ana-pantilie) [inline plutus-tx-plugin/test/Budget/9.6/andWithLocal.uplc.golden:10]: I was checking if this term and the "local" one are equivalent, and noticed something weird. It looks like we're overloading `lessThanInteger` here:
- on line 4 it's a bound variable
- on line 10 (outside ^'s scope) it's, I would guess, a builtin

I think the two terms (this and the "local" one) are alpha-equivalent, though.

File: plutus-tx-plugin/test/Budget/9.6/andWithoutGHCOpts.eval.golden [hunk 3]
```diff
@@ -0,0 +1,2 @@
+({cpu: 1822026
+| mem: 5504})
```
Comment (ana-pantilie) [inline plutus-tx-plugin/test/Budget/9.6/andWithoutGHCOpts.eval.golden:2]: This is equal to the "local" result. Shouldn't turning off all of the GHC optimisations have some impact as well?

File: plutus-tx-plugin/test/Budget/9.6/andWithLocal.uplc.golden (hunk 2 above)
Comment (zliu41) [reply plutus-tx-plugin/test/Budget/9.6/andWithLocal.uplc.golden:10]: They are not equivalent. The one introduced in line 4 corresponds to the one in `andWithLocal.pir.golden`, line 5. It returns the non-builtin `Bool` (which in PIR is just `Bool`, and in UPLC it is `constr 0 []` for True and `constr 1 []` for False). And it comes from `PlutusTx.Builtins.lessThanInteger`. In the `WithGhcOpts` test, it is inlined by GHC.

The `lessThanInteger` in line 10 is indeed the builtin function, and it returns the builtin `Bool`. The condition in the builtin `ifThenElse` requires the builtin `Bool`.

File: plutus-tx-plugin/test/Budget/9.6/andWithoutGHCOpts.eval.golden (hunk 3 above)
Comment (zliu41) [reply plutus-tx-plugin/test/Budget/9.6/andWithoutGHCOpts.eval.golden:2]: The PIRs are different but after the UPLC optimizations they become the same UPLC program.

The local one doesn't run much GHC optimization either. Even though the module is compiled with optimization, the plugin is installed quite early in the core-to-core pipeline, so it runs before most optimizations.

**Review by zliu41 — APPROVED** [review]
Looks great.

File: plutus-tx-plugin/test/Budget/9.6/andWithLocal.uplc.golden (hunk 2 above)
Comment (michaelpj) [reply plutus-tx-plugin/test/Budget/9.6/andWithLocal.uplc.golden:10]: Yes, there is name shadowing here, unfortunately.

File: plutus-tx-plugin/test/Budget/9.6/andWithoutGHCOpts.pir.golden [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+              {a -> a -> Bool}
+              (\(v : (\a -> a -> a -> Bool) a)
+                (v : a -> a -> Ordering)
+                (v : a -> a -> Bool)
+                (v : a -> a -> Bool)
+                (v : a -> a -> Bool)
+                (v : a -> a -> Bool)
+                (v : a -> a -> a)
+                (v : a -> a -> a) ->
+                 v)
```
[the remark this answers is not in the document]
Comment (michaelpj) [reply plutus-tx-plugin/test/Budget/9.6/andWithoutGHCOpts.pir.golden:75]: I think you mean specialization, but yeah, typeclass specialization is one of the complicated GHC optimizations that we would really benefit from but usually can't use.

File: plutus-tx-plugin/test/Budget/WithoutGHCOptimisations.hs [hunk 5]
```diff
@@ -0,0 +1,11 @@
+{-# OPTIONS_GHC
+ -O0 -fmax-simplifier-iterations=0
+ -fno-omit-interface-pragmas
+ -fno-ignore-interface-pragmas
+#-}
```
Comment (michaelpj) [inline plutus-tx-plugin/test/Budget/WithoutGHCOptimisations.hs:5]: nit for the future, we generally avoid multi-line pragmas, it's a bit less diff-friendly than doing one pragma per line with one option each

**Pull request merged**